src/KubernetesClient/Watcher.cs

```diff
@@ -174,6 +155,67 @@ private async Task WatcherLoop(CancellationToken cancellationToken)
             }
         }
 
+        internal static async IAsyncEnumerable<(WatchEventType, T)> CreateWatchEventEnumerator(
+            Func<Task<TextReader>> streamReaderCreator,
+            Action<Exception> onError = null,
+            [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            Task<TR> AttachCancellationToken<TR>(Task<TR> task)
+            {
+                if (!task.IsCompleted)
+                {
+                    // here to pass cancellationToken into task
+                    return task.ContinueWith(t => t.GetAwaiter().GetResult(), cancellationToken);
+                }
+
+                return task;
+            }
+
+            using var streamReader = await AttachCancellationToken(streamReaderCreator()).ConfigureAwait(false);
+
+            for (; ; )
+            {
+                // ReadLineAsync will return null when we've reached the end of the stream.
+                var line = await AttachCancellationToken(streamReader.ReadLineAsync()).ConfigureAwait(false);
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (line == null)
+                {
+                    yield break;
+                }
+
+                WatchEvent @event = null;
+
+                try
+                {
+                    var genericEvent = SafeJsonConvert.DeserializeObject<Watcher<KubernetesObject>.WatchEvent>(line);
+
+                    if (genericEvent.Object.Kind == "Status")
```

Comment:
@tg123, not sure, but if I'm not mistaken, if the list/watch operation actually returns existing objects, then those will be read (again) by the first ReadLineAsync above. The `line` will not deserialize properly, as the line represents the object list, not a watch event. Therefore a `NullReferenceException` will be thrown, caught below and reported via the `onError` handler if and only if an `onError` handler has been provided. Test cases do not cover this. I don't think this is intentional!?

The `LineSeparatedHttpContent` only peeks at the first line, so it is read again here. It doesn't know, whether it should discard the line as it doesn't know, whether it is reading an object list or a watch event.

Reply:
i noticed the npe during dev and did some changes to testcase https://github.com/kubernetes-client/csharp/pull/586/files/c8506d22c52caf3e827a51a9f4de477659557759#diff-b8e0f043f040cdeca7bd66470c6f4cf8dc42b39a1f514cf412e284309284e431L196

Yet, server will never return a `listobj` if `watch=true` and `LineSeparatedHttpContent` will install to `watch=true` only.

In the old days, (1.6 or 1.7 i believe), there is a `/watch/xxx` api which return `listobj` which will cause the issue you mentioned. but it was deprecated and `get` with `watch=true` will yield `watch obj` only
that is the reason i changed the testcase removing the first list obj.

Reply:
Then could you please have a look at the TestWatchWithHandlers? That's the test I stumbled on.

Reply:
yup
`await WriteStreamLine(httpContext, MockKubeApiServer.MockPodResponse).ConfigureAwait(false);`
should be removed 
it is outdated, but i did not notice it as test passed

Reply:
That is because there is no `onError` handler attached. I'm experimenting with an alternative implementation that encapsulates both success and error results in a `WatchEvent<T>`. That's why I noticed.